header.h

```diff
@@ -0,0 +1,167 @@
+#ifndef HEADER_H
+#define HEADER_H 
+
+#include <iostream>
+using std::cin;
+using std::cout;
+using std::endl;
+using std::cerr;
+
+using std::ostream;
+using std::istream;
+
+#include <cmath>
+using std::pow;
+using std::sqrt;
+
+#include <exception>
+using std::exception;
+
+#include <string>
+using std::string;
+
+#include <fstream>
+using std::ifstream;
+using std::ofstream;
+
+template <typename T, const unsigned int SIZE>
+class Point{
+public:
+
+	//constructor
+	Point();
+
+	//Calculate the distance between two points
+	double distance(const Point &) const;
+
+	//Overload > operator
+	bool operator>(const Point &) const;
+
+	//Overload stream insertion operator
+	ostream &toStream(ostream &os);
+
+	//Overload stream extraction operator
+	istream &fromStream(istream &is);
+
+	//Errors classes
+	class Error : public std::exception{
+	public:
+		Error(const char * e) : description(e) {};
+
+		const char * what() const {
+			return description;
+		}
+
+	private:
+		const char * description;
+	};
+
+
+	//Error for empty stream, nothing between ()
+	class EmptyStreamError : public Error{
+	public:
+		EmptyStreamError();
+		EmptyStreamError(const char * empty) : Error(empty) {};
+	};
+
+	//Error if the first or last element is not a ( or )
+	class InvalidSymbolError : public Error{
+	public:
+		InvalidSymbolError();
+		InvalidSymbolError(const char * invalid) : Error(invalid) {};
+	};
+private:
+
+	//The components of a point, can be 1D, 2D, 3D, etc.
+	T components[SIZE];
+};
+
+
+
+//Implementation
+//constructor, set every element of the components array to 0
+template <typename T, const unsigned int SIZE>
+Point<T, SIZE>::Point() : components{} {}
+
+//Distance function
+template <typename T, const unsigned int SIZE>
+double Point<T, SIZE>::distance(const Point &p) const{
+	T sum = 0;
+	for (int i = 0; i < SIZE; i++){
+		sum += pow((components[i] - p.components[i]), 2);
+	}
+
+	return sqrt(sum);
+}
+
+//Overloading greater than operator
+template <typename T, const unsigned int SIZE>
+bool Point<T, SIZE>::operator>(const Point &p) const{
+	return distance(Point{}) > p.distance(Point{});
+}
+
+//Overloading stream extraction operator
+template <typename T, const unsigned int SIZE>
+ostream &Point<T, SIZE>::toStream(ostream &os){
+	os << "( ";
+	for (int i = 0; i < SIZE; i++){
+		os << components[i] << " ";
+	}
+	os << " )";
+
+	return os;
+}
+
+//Overloading stream insertion operator
+template <typename T, const unsigned int SIZE>
+istream &Point<T, SIZE>::fromStream(istream &is) {
+	char parenthesis;
+	T temp;	//this variable will hold one component of a point and then pass it to components[]
+
+	is >> parenthesis;
+
+	//If nothing can be read (ex. empty file)
+	if (is.fail()){
+		is.clear();
+		throw Point<T,SIZE>::EmptyStreamError();
```

Comment:
i don't think the `Point<T,SIZE>::` is necessary here, because we're inside a method of `Point<T,SIZE>` and `EmptyStreamError` is declared inside the class template